### ISSUE TITLE
data fetched again on client when inside an async component

### ISSUE THREAD

**Issue by slarrasoain** [nuxt/nuxt#18405]
### Environment

------------------------------
- Operating System: `Windows_NT`
- Node Version:     `v18.12.1`
- Nuxt Version:     `3.0.0`
- Nitro Version:    `1.0.0`
- Package Manager:  `npm@8.19.2`
- Builder:          `vite`
- User Config:      `-`
- Runtime Modules:  `-`
- Build Modules:    `-`
------------------------------

### Reproduction

https://stackblitz.com/edit/nuxt-starter-jyws2q

### Describe the bug

In my nuxt3 application (CMS) i notice that sometime  query are run twice on SSR
I make a simple repro with the structure of the application

When query are run in deep level module loaded dynamicaly useAsyncData of the module run twice 
Two query are simulated the pageData query in page and todoQuery in last component 'Block1'

I made 3 test pages in repro 
- index.vue is like in my application  => the todo query is run twice [BUG]
- dynamic.vue skip PageContent component and call TplPage component dynamicaly => the todo query is run twice [BUG]
- direct.vue skip PageContent component and call TplPage component directly  => the todo query is run once

**Comment by m7rlin** [comment]
I have same problem.

**Comment by mercs600** [comment]
We are facing the same issue, components rendered by `component :is` call useAsyncData twice - once on SSR and once on client side. Hydrating flag is setup on false and there is no data in  nuxt.static.data 

```js
  const getCachedData = () => nuxt.isHydrating ? nuxt.payload.data[key] : nuxt.static.data[key]
```

When it comes to useAsyncData it should not work this way

**Comment by mercs600** [comment]
I have found workaround for dynamic components, but I'm not sure how safe it is :) 
```vue
<script setup lang="ts">
useNuxtApp().isHydrating = true
const done = useNuxtApp().deferHydration()

const { data } = await useAsyncData('test', async () => {
  return await $fetch('https://fakestoreapi.com/products')
})

onMounted(async () => {
  await done()
})
</script>
```

**Comment by zhanghd0824** [comment]
In your example, I found that the index query is triggered on both the server and client, while the todo query is requested once on the server and twice on the client

My question is: useAsyncData will request once in both server and client, is this normal? I was originally doing it for SEO, the server has already sorted out the dom, why does the client have to request it once? @danielroe

**Comment by danielroe** [comment]
Looking into this, this is caused by an upstream Vue bug with Suspense + async components: https://github.com/vuejs/core/issues/6638. The workaround provided by @mercs600 is fine for now, though I think you can call `done()` directly after the `useAsyncData` call, and not just in mounted.

I cannot reproduce the issue with the index route, so I think it's been resolved since the reproduction.

**Comment by danielroe** [comment]
To clarify, the problem is that those (global) components are loaded async, which triggers the bug. You can directly import them to prevent this, e.g.:

```ts
import { TplPage } from '#components'
```

**Comment by slarrasoain** [comment]
> on is: useAsyncData will request once in both server and client, is this normal? I was originally doing it for SEO, the server has already sorted out the dom, why does the client have to request it once? @danielroe

In my actual case, I can't import the component directly because I've built a cms with a block system and the blocks used in the rendering page aren't known in advance, hence the use of global components. Will try the workaround thank a lot.

**Comment by mkarras** [comment]
The requests are also incorrectly called on client using Nuxt 3.6.3 for a default SSR page load.
We fetch data in parallel with:
```
 const [
    { data: yData, error: xError },
    { data: yData, error: yError },
    ...
  ] = await Promise.all([
    useAsyncData(getLocalizedKey('xData'), () => fetchXData()),
    useAsyncData(getLocalizedKey('yData'), () => fetchYData()),
    ...
  ]);
```
Works well with Nuxt 3.5.3.
The workaround from @mercs600 works.

Any progress here solving the issue @danielroe ?
Should i create a new ticket because its related to any page load now starting from Nuxt 3.6.x (at least 3.6.3)?

**Comment by danielroe** [comment]
This is not a new issue. The root cause is https://github.com/vuejs/core/issues/6638 and the workaround is above: https://github.com/nuxt/nuxt/issues/18405#issuecomment-1548012719.

**Comment by mkarras** [comment]
@danielroe but im wondering why it works with Nuxt 3.5.3 but for 3.6.3 not. https://github.com/vuejs/core/issues/6638 is also present for Nuxt 3.5.3.

**Comment by mkarras** [comment]
@danielroe updated to Nuxt 3.7.2 and the issue "useAsyncData is called another time on client when a page is server side rendered" introduced with Nuxt 3.6.3 is gone.

The problem now is, that Nuxt 3.7.2 does not work in dev mode. Only in prod.
```
...
FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
...
```
But this is a different issue. I'll see if there are already issues like this and potentially post a new one. Or do you have a guess as to what it potentially came in with? Already came in with Nuxt 3.7.

**Comment by codeflorist** [comment]
This is also a possible workaround:

```js
const { data } = await useAsyncData('test', async () => {
  return await $fetch('https://fakestoreapi.com/products')
},
{
  getCachedData(key, nuxt) {
    return nuxt.payload.data[key] || nuxt.static.data[key]
  },
})
```

The default of `getCachedData` is:

```js
  getCachedData(key, nuxt) {
    return nuxt.isHydrating ? nuxt.payload.data[key] : nuxt.static.data[key]
  },
```

@danielroe 
Since this issue is now open for 1.5 years: Would simply changing the default of `getCachedData` be a viable fix - at least until the vue-bug is fixed?

**Comment by danielroe** [comment]
I would be open to that, if we have a good solution that's consistent also with client-side navigation.

**Comment by codeflorist** [comment]
> I would be open to that, if we have a good solution that's consistent also with client-side navigation.

So this is problematic in your opinion? ->

```js
  getCachedData(key, nuxt) {
    return nuxt.payload.data[key] || nuxt.static.data[key]
  },
```

So far it seems to be working ok on my side - also with client-side navigation.

**Comment by dargmuesli** [comment]
You might want to try Vue v`3.5.0-alpha.4` which should include the upsteam fix.

**Comment by simonmaass** [comment]
i have the same problem.. i have a page that has a component :is setup and then in the components i fetch data (in my case i check if it is a product or category url and the set the according component.

In this case the useAsnycFetch is called twice on client side...
